File: src/main/java/me/alegian/thavma/impl/common/menu/slot/SlotRange.java
package me.alegian.thavma.impl.common.menu.slot;

import me.alegian.thavma.impl.common.menu.Menu;

/**
 * Keeps track of slot index ranges. Need to call start() and end() before the
 * addition of the first slot and after the addition of the last slot
 */
public class SlotRange {
  protected Menu menu;
  protected int start;
  protected int end;

  public SlotRange(Menu menu) {
    this.menu = menu;
    this.start = 0;
    this.end = 0;
  }

  public void start() {
    this.start = this.menu.slots.size();
  }

  public void end() {
    this.end = this.menu.slots.size() - 1;
  }

  public int getStart() {
    return this.start;
  }

  public int getEnd() {
    return this.end;
  }

  public boolean contains(int slotId) {
    return slotId >= this.start && slotId <= this.end;
  }

  /**
   * Tracking single slots should be done AFTER adding them
   */
  public static class Single extends SlotRange {
    public Single(Menu menu) {
      super(menu);
    }

    public void track() {
      this.start = this.menu.slots.size() - 1;
      this.end = this.menu.slots.size() - 1;
    }

    public boolean is(int slotId) {
      return this.contains(slotId);
    }
  }
}

File: src/main/java/me/alegian/thavma/impl/common/menu/slot/SlotRange.kt
package me.alegian.thavma.impl.common.menu.slot

import me.alegian.thavma.impl.common.menu.Menu
import net.minecraft.world.inventory.Slot

/**
 * Keeps track of slot index ranges. Need to call start() and end() before the
 * addition of the first slot and after the addition of the last slot
 */
open class SlotRange(protected var menu: Menu) {
  var start: Int = 0
    protected set
  var end: Int = 0
    protected set

  fun start() {
    this.start = menu.slots.size
  }

  fun end() {
    this.end = menu.slots.size - 1
  }

  fun contains(slotId: Int): Boolean {
    return slotId >= this.start && slotId <= this.end
  }

  val slots: List<Slot>
    get() = menu.slots.subList(this.start, this.end + 1)

  /**
   * Tracking single slots should be done AFTER adding them
   */
  class Single(menu: Menu) : SlotRange(menu) {
    fun track() {
      this.start = menu.slots.size - 1
      this.end = menu.slots.size - 1
    }

    val slot: Slot
      get() = menu.slots[start]
  }
}
